csrc/device_lower/pass/expr_sort.cpp

```diff
@@ -593,60 +587,43 @@ void ExprGroup::clearTraversalInfo() {
   payload()->merged = false;
 }
 
-void ExprSegmentationSorter::resetTraversal() {
-  for (auto& group : groups_) {
-    // Start traversal at input groups
-    if (group->producerEdges().empty()) {
-      to_visit_.push_back(group.get());
-    }
-    group->clearTraversalInfo();
-  }
-}
-
 // Level is maximum distance from inputs. It's the metric used to select what
 // nodes can be merged while maintaining a DAG
 void ExprSegmentationSorter::resetLevels() {
-  std::vector<ExprGroup*> next_to_visit;
-
-  while (!to_visit_.empty()) {
-    auto visit = to_visit_.front();
-    to_visit_.pop_front();
-
-    // All inputs processed?
-    bool ready = true;
-    if (!visit->producerEdges().empty()) {
-      ready = std::all_of(
-          visit->producerEdges().begin(),
-          visit->producerEdges().end(),
-          [&](ExprGroupConnections* dep) {
-            return dep->from->payload()->visited;
-          });
-    }
-
-    if (!ready) {
-      // In case traversal doesn't complete because there's an error in the
-      // DAG topology.
-      next_to_visit.push_back(visit);
-      continue;
-    }
-
-    visit->payload()->visited = true;
-
-    to_visit_.insert(
-        to_visit_.end(), next_to_visit.begin(), next_to_visit.end());
-    next_to_visit.clear();
+  std::deque<ExprGroup*> to_visit;
```

Comment:
This can be merged with segmentation through some template-ing, but I haven't attempted it in this PR to keep the scope limited.